# Import Modules
---

In [1]:
import pandas as pd
import numpy as np
import requests
import tweepy
import json
import os

from tqdm.notebook import tqdm

<a id='toc'></a>
# Table of Contents
---

* [I. Gathering](#gathering)
* [II. Assessing](#assessing)
* [III. Cleaning](#cleaning)
* [IV. Analyzing](#analyzing)
* [III. Visualizing](#visualizing)

<a id='gathering'></a>
# I. Gathering Data 
---

In [2]:
data_dir = "./data"

### a. WeRateDogs Twitter Archive

In [3]:
wrd_data_filepath = 'twitter-archive-enhanced.csv'

df_wrd_twitter_old = pd.read_csv(os.path.join(data_dir, wrd_data_filepath))

### b. Tweet Image Predictions

In [4]:
image_predictions_filepath = os.path.join(data_dir, 'image-predictions.tsv')

In [5]:
'''
Uncomment the following code to download data if file(image-predictions.tsv) does not exist
'''

# url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'

# response = requests.get(url, allow_redirects=True)

# with open(image_predictions_filepath, 'wb') as f:
#     f.write(response.content)

'\nUncomment the following code to download data if file(image-predictions.tsv) does not exist\n'

In [6]:
df_image_predictions_old = pd.read_csv(image_predictions_filepath, sep="\t")

### c. Any Additional Data

**Via Twitter API**

In [7]:
json_filepath = os.path.join(data_dir, "tweet_json.txt")

In [8]:
'''
Uncomment the following code to download data if file(tweet_json.txt) does not exist
'''

# with open('twitter-credential.json') as f:
#     credentials = json.load(f)

'\nUncomment the following code to download data if file(tweet_json.txt) does not exist\n'

In [9]:
'''
Uncomment the following code to download data if file(tweet_json.txt) does not exist
'''

# auth = tweepy.OAuthHandler(credentials['consumer_key'], credentials['consumer_secret'])

# auth.set_access_token(credentials['access_token'], credentials['access_token_secret'])

# api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

'\nUncomment the following code to download data if file(tweet_json.txt) does not exist\n'

In [10]:
'''
Uncomment the following code to download data if file(tweet_json.txt) does not exist
'''

# tweet_ids = df_wrd_twitter_old['tweet_id']

# statuses = []

# for tweet_id in tqdm(tweet_ids):

#     try:

#         status = api.get_status(tweet_id, tweet_mode='extended')

#         statuses.append(status._json)

#     except:

#         continue

'\nUncomment the following code to download data if file(tweet_json.txt) does not exist\n'

In [11]:
'''
Uncomment the following code to download data if file(tweet_json.txt) does not exist
'''

# with open(json_filepath, 'w') as outfile:
#     for status in statuses:
#         json.dump(status, outfile)
#         outfile.write("\n")

'\nUncomment the following code to download data if file(tweet_json.txt) does not exist\n'

In [12]:
df_tweet_json_old = pd.read_json(json_filepath, lines=True)

**Via Kaggle** [For all available dog breeds]

Reference: [https://www.kaggle.com/c/dog-breed-identification/data](https://www.kaggle.com/c/dog-breed-identification/data)

In [13]:
df_dog_breeds_old = pd.read_csv(os.path.join(data_dir, "labels.csv"))

<a id='assessing'></a>
# II. Assessing Data
---

In [14]:
df_wrd_twitter_old

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

In [15]:
df_wrd_twitter_old.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [16]:
"None" in df_wrd_twitter_old.values

True

In [17]:
df_wrd_twitter_old.isin(["None"]).sum()

tweet_id                         0
in_reply_to_status_id            0
in_reply_to_user_id              0
timestamp                        0
source                           0
text                             0
retweeted_status_id              0
retweeted_status_user_id         0
retweeted_status_timestamp       0
expanded_urls                    0
rating_numerator                 0
rating_denominator               0
name                           745
doggo                         2259
floofer                       2346
pupper                        2099
puppo                         2326
dtype: int64

In [18]:
df_wrd_twitter_old.query("doggo != 'None' and floofer != 'None'")['tweet_id'].count()

1

In [19]:
df_wrd_twitter_old.query("doggo != 'None' and pupper != 'None'")['tweet_id'].count()

12

In [20]:
df_wrd_twitter_old.query("doggo != 'None' and puppo != 'None'")['tweet_id'].count()

1

In [21]:
df_wrd_twitter_old.query("doggo == 'None' and floofer == 'None' and pupper == 'None' and puppo == 'None' and name == 'None'")['tweet_id'].count()

585

In [22]:
df_wrd_twitter_old['rating_denominator'].unique()

array([ 10,   0,  15,  70,   7,  11, 150, 170,  20,  50,  90,  80,  40,
       130, 110,  16, 120,   2])

In [23]:
df_wrd_twitter_old.query("rating_denominator == 0")['rating_numerator'].unique()

array([960])

In [24]:
df_wrd_twitter_old.query("rating_denominator == 0").index

Int64Index([313], dtype='int64')

In [25]:
df_wrd_twitter_old['name'].unique()

array(['Phineas', 'Tilly', 'Archie', 'Darla', 'Franklin', 'None', 'Jax',
       'Zoey', 'Cassie', 'Koda', 'Bruno', 'Ted', 'Stuart', 'Oliver',
       'Jim', 'Zeke', 'Ralphus', 'Canela', 'Gerald', 'Jeffrey', 'such',
       'Maya', 'Mingus', 'Derek', 'Roscoe', 'Waffles', 'Jimbo', 'Maisey',
       'Lilly', 'Earl', 'Lola', 'Kevin', 'Yogi', 'Noah', 'Bella',
       'Grizzwald', 'Rusty', 'Gus', 'Stanley', 'Alfy', 'Koko', 'Rey',
       'Gary', 'a', 'Elliot', 'Louis', 'Jesse', 'Romeo', 'Bailey',
       'Duddles', 'Jack', 'Emmy', 'Steven', 'Beau', 'Snoopy', 'Shadow',
       'Terrance', 'Aja', 'Penny', 'Dante', 'Nelly', 'Ginger', 'Benedict',
       'Venti', 'Goose', 'Nugget', 'Cash', 'Coco', 'Jed', 'Sebastian',
       'Walter', 'Sierra', 'Monkey', 'Harry', 'Kody', 'Lassie', 'Rover',
       'Napolean', 'Dawn', 'Boomer', 'Cody', 'Rumble', 'Clifford',
       'quite', 'Dewey', 'Scout', 'Gizmo', 'Cooper', 'Harold', 'Shikha',
       'Jamesy', 'Lili', 'Sammy', 'Meatball', 'Paisley', 'Albus',
       'Nept

In [26]:
df_wrd_twitter_old[df_wrd_twitter_old['name'].str.islower()]['name'].unique()

array(['such', 'a', 'quite', 'not', 'one', 'incredibly', 'mad', 'an',
       'very', 'just', 'my', 'his', 'actually', 'getting', 'this',
       'unacceptable', 'all', 'old', 'infuriating', 'the', 'by',
       'officially', 'life', 'light', 'space'], dtype=object)

In [27]:
df_image_predictions_old

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

In [28]:
df_image_predictions_old.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [29]:
df_image_predictions_old['p1'].unique()

array(['Welsh_springer_spaniel', 'redbone', 'German_shepherd',
       'Rhodesian_ridgeback', 'miniature_pinscher',
       'Bernese_mountain_dog', 'box_turtle', 'chow', 'shopping_cart',
       'miniature_poodle', 'golden_retriever', 'Gordon_setter',
       'Walker_hound', 'pug', 'bloodhound', 'Lhasa', 'English_setter',
       'hen', 'desktop_computer', 'Italian_greyhound', 'Maltese_dog',
       'three-toed_sloth', 'ox', 'malamute', 'guinea_pig',
       'soft-coated_wheaten_terrier', 'Chihuahua',
       'black-and-tan_coonhound', 'coho', 'toy_terrier',
       'Blenheim_spaniel', 'Pembroke', 'llama',
       'Chesapeake_Bay_retriever', 'curly-coated_retriever', 'dalmatian',
       'Ibizan_hound', 'Border_collie', 'Labrador_retriever', 'seat_belt',
       'snail', 'miniature_schnauzer', 'Airedale', 'triceratops', 'swab',
       'hay', 'hyena', 'jigsaw_puzzle', 'West_Highland_white_terrier',
       'toy_poodle', 'giant_schnauzer', 'vizsla', 'vacuum', 'Rottweiler',
       'Siberian_husky', 't

In [30]:
df_image_predictions_old['p2'].unique()

array(['collie', 'miniature_pinscher', 'malinois', 'redbone',
       'Rottweiler', 'English_springer', 'mud_turtle', 'Tibetan_mastiff',
       'shopping_basket', 'komondor', 'Yorkshire_terrier',
       'English_foxhound', 'bull_mastiff', 'German_shepherd', 'Shih-Tzu',
       'Newfoundland', 'cock', 'desk', 'toy_terrier', 'toy_poodle',
       'otter', 'Chesapeake_Bay_retriever', 'Siberian_husky', 'skunk',
       'Afghan_hound', 'bloodhound', 'barracouta', 'papillon',
       'cocker_spaniel', 'chow', 'Irish_terrier', 'chain_saw', 'beagle',
       'giant_schnauzer', 'Labrador_retriever', 'Pembroke', 'Chihuahua',
       'Weimaraner', 'slug', 'Brittany_spaniel', 'standard_schnauzer',
       'teddy', 'armadillo', 'African_hunting_dog', 'vizsla', 'doormat',
       'pug', 'Italian_greyhound', 'Samoyed', 'Pomeranian',
       'miniature_poodle', 'Lakeland_terrier', 'Irish_setter', 'swab',
       'malamute', 'bath_towel', 'Border_collie', 'Leonberg', 'drake',
       'French_bulldog', 'ice_bear', 

In [31]:
df_image_predictions_old['p3'].unique()

array(['Shetland_sheepdog', 'Rhodesian_ridgeback', 'bloodhound',
       'miniature_pinscher', 'Doberman', 'Greater_Swiss_Mountain_dog',
       'terrapin', 'fur_coat', 'golden_retriever',
       'soft-coated_wheaten_terrier', 'Labrador_retriever', 'Pekinese',
       'Ibizan_hound', 'French_bulldog', 'malinois', 'Dandie_Dinmont',
       'borzoi', 'partridge', 'bookcase', 'basenji', 'miniature_poodle',
       'great_grey_owl', 'groenendael', 'Eskimo_dog', 'hamster', 'briard',
       'papillon', 'flat-coated_retriever', 'gar', 'Chihuahua',
       'Shih-Tzu', 'Pomeranian', 'dingo', 'power_drill', 'Saluki',
       'Great_Pyrenees', 'West_Highland_white_terrier', 'collie',
       'toy_poodle', 'vizsla', 'acorn', 'giant_schnauzer', 'teddy',
       'common_iguana', 'wig', 'water_buffalo', 'coyote', 'seat_belt',
       'kelpie', 'space_heater', 'Brabancon_griffon', 'standard_poodle',
       'beagle', 'Irish_water_spaniel', 'bluetick', 'Weimaraner',
       'Chesapeake_Bay_retriever', 'toilet_tiss

In [32]:
df_tweet_json_old

created_at                  id              id_str  \
0    2017-08-01 16:23:56+00:00  892420643555336193  892420643555336192   
1    2017-08-01 00:17:27+00:00  892177421306343426  892177421306343424   
2    2017-07-31 00:18:03+00:00  891815181378084864  891815181378084864   
3    2017-07-30 15:58:51+00:00  891689557279858688  891689557279858688   
4    2017-07-29 16:00:24+00:00  891327558926688256  891327558926688256   
...                        ...                 ...                 ...   
2326 2015-11-16 00:24:50+00:00  666049248165822465  666049248165822464   
2327 2015-11-16 00:04:52+00:00  666044226329800704  666044226329800704   
2328 2015-11-15 23:21:54+00:00  666033412701032449  666033412701032448   
2329 2015-11-15 23:05:30+00:00  666029285002620928  666029285002620928   
2330 2015-11-15 22:32:08+00:00  666020888022790149  666020888022790144   

                                              full_text  truncated  \
0     This is Phineas. He's a mystical boy. Only eve...      False   
1     This is Tilly. She's just checking pup on you....      False   
2     This is Archie. He is a rare Norwegian Pouncin...      False   
3     This is Darla. She commenced a snooze mid meal...      False   
4     This is Franklin. He would like you to stop ca...      False   
...                                                 ...        ...   
2326  Here we have a 1949 1st generation vulpix. Enj...      False   
2327  This is a purebred Piers Morgan. Loves to Netf...      False   
2328  Here is a very happy pup. Big fan of well-main...      False   
2329  This is a western brown Mitsubishi terrier. Up...      False   
2330  Here we have a Japanese Irish Setter. Lost eye...      False   

     display_text_range                                           entities  \
0               [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1              [0, 138]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2              [0, 121]  {'hashtags': [], 'symbols': [], 'user_mentions...   
3               [0, 79]  {'hashtags': [], 'symbols': [], 'user_mentions...   
4              [0, 138]  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   
...                 ...                                                ...   
2326           [0, 120]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2327           [0, 137]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2328           [0, 130]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2329           [0, 139]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2330           [0, 131]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  \
0     {'media': [{'id': 892420639486877696, 'id_str'...   
1     {'media': [{'id': 892177413194625024, 'id_str'...   
2     {'media': [{'id': 891815175371796480, 'id_str'...   
3     {'media': [{'id': 891689552724799489, 'id_str'...   
4     {'media': [{'id': 891327551943041024, 'id_str'...   
...                                                 ...   
2326  {'media': [{'id': 666049244999131136, 'id_str'...   
2327  {'media': [{'id': 666044217047650304, 'id_str'...   
2328  {'media': [{'id': 666033409081393153, 'id_str'...   
2329  {'media': [{'id': 666029276303482880, 'id_str'...   
2330  {'media': [{'id': 666020881337073664, 'id_str'...   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2326  <a href="http://twitter.com/download/iphone" r...   
2327  <a href="http://twitter.com/download/iphone" r...   
2328  <a href="http://twitter.com/download/iphone" r...   
2329  <a href="http://twitter.com/download/iphone" r..

In [33]:
df_tweet_json_old.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 32 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   created_at                     2331 non-null   datetime64[ns, UTC]
 1   id                             2331 non-null   int64              
 2   id_str                         2331 non-null   int64              
 3   full_text                      2331 non-null   object             
 4   truncated                      2331 non-null   bool               
 5   display_text_range             2331 non-null   object             
 6   entities                       2331 non-null   object             
 7   extended_entities              2059 non-null   object             
 8   source                         2331 non-null   object             
 9   in_reply_to_status_id          77 non-null     float64            
 10  in_reply_to_status_id_st

In [34]:
df_tweet_json_old.query('retweeted == True')['id'].nunique()

0

In [35]:
df_tweet_json_old.query("retweet_count > 0")['id'].nunique()

2331

In [36]:
df_tweet_json_old.query('favorited == True')['id'].nunique()

0

In [37]:
df_tweet_json_old.query("favorite_count > 0")['id'].nunique()

2168

In [38]:
df_dog_breeds_old.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10222 entries, 0 to 10221
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      10222 non-null  object
 1   breed   10222 non-null  object
dtypes: object(2)
memory usage: 159.8+ KB


In [39]:
df_dog_breeds_old['breed'].unique()

array(['boston_bull', 'dingo', 'pekinese', 'bluetick', 'golden_retriever',
       'bedlington_terrier', 'borzoi', 'basenji', 'scottish_deerhound',
       'shetland_sheepdog', 'walker_hound', 'maltese_dog',
       'norfolk_terrier', 'african_hunting_dog',
       'wire-haired_fox_terrier', 'redbone', 'lakeland_terrier', 'boxer',
       'doberman', 'otterhound', 'standard_schnauzer',
       'irish_water_spaniel', 'black-and-tan_coonhound', 'cairn',
       'affenpinscher', 'labrador_retriever', 'ibizan_hound',
       'english_setter', 'weimaraner', 'giant_schnauzer', 'groenendael',
       'dhole', 'toy_poodle', 'border_terrier', 'tibetan_terrier',
       'norwegian_elkhound', 'shih-tzu', 'irish_terrier', 'kuvasz',
       'german_shepherd', 'greater_swiss_mountain_dog', 'basset',
       'australian_terrier', 'schipperke', 'rhodesian_ridgeback',
       'irish_setter', 'appenzeller', 'bloodhound', 'samoyed',
       'miniature_schnauzer', 'brittany_spaniel', 'kelpie', 'papillon',
       'borde

> **Twitter Reference**:
>
> * Tweet objects: https://developer.twitter.com/en/docs/tweets/data-dictionary/overview/tweet-object

> **Key Points:**
> * Only original ratings (no retweets) that have images is wanted
> * The tweets beyond August 1st, 2017 do not need to be gathered.
> * Cleaning includes merging individual pieces of data according to the rules of tidy data
> * The rating numerators are greater than the denominators does not need to be cleaned

> **Quality Issues:**
> * ~~IDs are found to have wrong types~~
> * ~~Data contains retweets~~
> * ~~Variable **retweeted** indicates whether this Tweet has been Retweeted by the authenticating user. Variable **retweet_count** indicates number of times this Tweet has been retweeted. The maxium of retweet_count reaches 77154 but the only possible value for retweeted is False~~
> * ~~The above issue is also found between **favourited** and **favorite_count**~~
> * Misleading column names in **df_image_predictions_old**: **p1**, **p1_conf**, **p1_dog**, **p2**, **p2_conf**, **p2_dog**, **p3**, **p3_conf**, **p3_dog**
> * Column name **text** is not descriptive
> * ~~Issues with columns **p1**, **p2**, **p3**~~
>  ~~* Not all words start with capitalized characters~~
>  ~~* Some uses "_", and some uses "-"~~
> * ~~**timestamp** is not of type DateTime.~~ Same as **retweeted_status_timestamp**
> * ~~The **name** column in **df_wrd_twitter_old** contains words, including "None", other than dog names. Same as **p1**, **p2**, **p3** in **df_image_predictions_old**~~
> * ~~The **rating_denominator** contains 0, but its corresponding **rating_numerator** is 960~~
> * ~~There exist some entries not having dog "states". In the meantime, there exist some entries having two dog "states"~~
> * ~~In **df_dog_breeds_old**, Some names contains "-" while some uses "_"~~
> * ~~There exist duplicates in the **breed** column in **df_dog_breeds_old**~~
> * ~~Some information are shown in 2 columns in **df_tweet_json_old** for different data types~~

> **Tidiness Issues:**
> * Dog "states" are seperated into 4 columns: **doggo**, **floofer**, **pupper** and **puppo**  
> * Algorithms used are seperated into 3 columns: **p1**, **p2**, **p3**
> * Tweet information and dog information in the same table

<a id='cleaning'></a>
# III. Cleaning Data 
---

In [40]:
df_dog_breeds = df_dog_breeds_old.copy()
df_tweet_json = df_tweet_json_old.copy()
df_image_predictions = df_image_predictions_old.copy()
df_wrd_twitter = df_wrd_twitter_old.copy()

---
### Define

* Remove duplicates in breed column in df_dog_breeds using `drop_duplicates`
* Replace - with _ in breed column using `str.replace`

### Code

In [41]:
df_dog_breeds['breed'].str.contains('-').sum(), df_dog_breeds.duplicated(['breed']).any()

(561, True)

In [42]:
df_dog_breeds.drop_duplicates(['breed'], inplace=True)

df_dog_breeds['breed'] = df_dog_breeds['breed'].str.replace('-', '_')

### Test

In [43]:
df_dog_breeds['breed'].str.contains('-').sum(), df_dog_breeds.duplicated(['breed']).any()

(0, False)

---
### Define

* Remove columns end with str in **df_tweet_json** using `drop`
* Change data types of all id columns in **df_tweet_json** to str using `astype` 

### Code

In [44]:
cols_to_drop = [col for col in df_tweet_json if col.endswith('str')]

len(cols_to_drop)

4

In [45]:
cols_to_change_type = [col for col in df_tweet_json if col.endswith('id')]

len(cols_to_change_type)

4

In [46]:
for col in cols_to_change_type:
     print(col, df_tweet_json[col].dtype)

id int64
in_reply_to_status_id float64
in_reply_to_user_id float64
quoted_status_id float64


In [47]:
df_tweet_json.drop(columns=cols_to_drop, inplace=True)

In [48]:
for col in cols_to_change_type:
    df_tweet_json[col] = df_tweet_json[col].astype(str)

### Test

In [49]:
cols_to_drop = [col for col in df_tweet_json if col.endswith('str')]

len(cols_to_drop)

0

In [50]:
for col in cols_to_change_type:
     print(col, df_tweet_json[col].dtype)

id object
in_reply_to_status_id object
in_reply_to_user_id object
quoted_status_id object


---
### Define

Remove **retweeted** and **favourited** columns in **df_tweet_json** using `drop`

### Code

In [51]:
cols = ['retweeted', 'favorited']

cols_to_drop = [col for col in df_tweet_json if col in cols]

len(cols_to_drop)

2

In [52]:
df_tweet_json.drop(columns=cols_to_drop, inplace=True)

### Test

In [53]:
cols_to_drop = [col for col in df_tweet_json if col in cols]

len(cols_to_drop)

0

---
### Define

Filter out retweets in **df_tweet_json** by checking if **retweeted_status** is null

### Code

In [54]:
df_tweet_json['retweeted_status'].notnull().any()

True

In [55]:
row_indices_to_remove = df_tweet_json.index[df_tweet_json['retweeted_status'].notnull()].tolist()

In [56]:
df_tweet_json.drop(row_indices_to_remove, inplace=True)

### Test

In [57]:
df_tweet_json['retweeted_status'].notnull().any()

False

---
### Define

Change the data type of **tweet_id** to object using `astype`

### Code

In [58]:
df_image_predictions['tweet_id'].dtype

dtype('int64')

In [59]:
df_image_predictions['tweet_id'] = df_image_predictions['tweet_id'].astype(str)

### Test

In [60]:
df_image_predictions['tweet_id'].dtype

dtype('O')

---
### Define

* Fix format issues in **p1**, **p2** and **p3**. Final format: lowercase with "_" as seperator
* Filter out rows with nonsense dog breeds

### Code

In [61]:
p1_sum = df_image_predictions['p1'].str.contains('-').sum()
p2_sum = df_image_predictions['p2'].str.contains('-').sum()
p3_sum = df_image_predictions['p3'].str.contains('-').sum()

p1_sum, p2_sum, p3_sum

(59, 66, 61)

In [62]:
df_image_predictions['p1'] = df_image_predictions['p1'].str.lower()
df_image_predictions['p2'] = df_image_predictions['p2'].str.lower()
df_image_predictions['p3'] = df_image_predictions['p3'].str.lower()

df_image_predictions['p1'] = df_image_predictions['p1'].str.replace('-', '_')
df_image_predictions['p2'] = df_image_predictions['p2'].str.replace('-', '_')
df_image_predictions['p3'] = df_image_predictions['p3'].str.replace('-', '_')

In [63]:
'''
Filter out nonsense names
'''

index_p1 = df_image_predictions[~df_image_predictions.p1.isin(df_dog_breeds['breed'])].index
df_image_predictions.drop(index_p1, inplace=True)

index_p2 = df_image_predictions[~df_image_predictions.p2.isin(df_dog_breeds['breed'])].index
df_image_predictions.drop(index_p2, inplace=True)

index_p3 = df_image_predictions[~df_image_predictions.p3.isin(df_dog_breeds['breed'])].index
df_image_predictions.drop(index_p3, inplace=True)

### Test

In [64]:
df_image_predictions['p1'].str.contains('-').sum()

0

In [65]:
df_image_predictions['p2'].str.contains('-').sum()

0

In [66]:
df_image_predictions['p3'].str.contains('-').sum()

0

In [67]:
index_p1 = df_image_predictions[~df_image_predictions.p1.isin(df_dog_breeds['breed'])].index
index_p2 = df_image_predictions[~df_image_predictions.p2.isin(df_dog_breeds['breed'])].index
index_p3 = df_image_predictions[~df_image_predictions.p3.isin(df_dog_breeds['breed'])].index

index_p1.shape[0], index_p2.shape[0], index_p3.shape[0]

(0, 0, 0)

### Define

* Filter out retweets by removing the ones with not null **retweeted_status_id**
* Filter out rows with **rating_denominator** = 0
* Change the data type of **tweet_id** from number to string type
* Change the data type of **timestamp** from object to timestamp
* Filter out rows with no dog states or more than one dog state
* Filter out dog name "None"

### Code

In [68]:
index_to_remove = df_wrd_twitter[df_wrd_twitter['retweeted_status_id'].notnull()].index

df_wrd_twitter.drop(index_to_remove, inplace=True)

In [69]:
index_to_remove = df_wrd_twitter.query("rating_denominator == 0").index

df_wrd_twitter.drop(index_to_remove, inplace=True)

In [70]:
df_wrd_twitter['tweet_id'] = df_wrd_twitter['tweet_id'].astype(str)

In [71]:
df_wrd_twitter['timestamp'] = pd.to_datetime(df_wrd_twitter['timestamp'])

In [72]:
condition_1 = "(doggo=='doggo' and floofer=='floofer')"
condition_2 = "(doggo=='doggo' and pupper=='pupper')"
condition_3 = "(doggo=='doggo' and puppo=='puppo')"
condition_4 = "(floofer=='floofer' and pupper=='pupper')"
condition_5 = "(floofer=='floofer' and puppo=='puppo')"
condition_6 = "(pupper=='pupper' and puppo=='puppo')"
condition_7 = "(doggo=='None' and floofer=='None' and pupper=='None' and puppo=='puppo')"

arr = [condition_1, condition_2, condition_3, condition_4, condition_5, condition_6, condition_7]

condition = " or ".join(arr)

rows_to_remove = df_wrd_twitter.query(condition)

df_wrd_twitter.drop(rows_to_remove.index, inplace=True)

In [73]:
indices_to_remove = df_wrd_twitter.query("name=='None'").index

df_wrd_twitter.drop(indices_to_remove, inplace=True)

### Test

In [74]:
df_wrd_twitter.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1474 entries, 0 to 2354
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    1474 non-null   object             
 1   in_reply_to_status_id       1 non-null      float64            
 2   in_reply_to_user_id         1 non-null      float64            
 3   timestamp                   1474 non-null   datetime64[ns, UTC]
 4   source                      1474 non-null   object             
 5   text                        1474 non-null   object             
 6   retweeted_status_id         0 non-null      float64            
 7   retweeted_status_user_id    0 non-null      float64            
 8   retweeted_status_timestamp  0 non-null      object             
 9   expanded_urls               1474 non-null   object             
 10  rating_numerator            1474 non-null   int64           

In [75]:
df_wrd_twitter['retweeted_status_id'].unique()

array([nan])

In [76]:
df_wrd_twitter.query("rating_denominator == 0").index

Int64Index([], dtype='int64')

In [77]:
condition_1 = "(doggo=='doggo' and floofer=='floofer')"
condition_2 = "(doggo=='doggo' and pupper=='pupper')"
condition_3 = "(doggo=='doggo' and puppo=='puppo')"
condition_4 = "(floofer=='floofer' and pupper=='pupper')"
condition_5 = "(floofer=='floofer' and puppo=='puppo')"
condition_6 = "(pupper=='pupper' and puppo=='puppo')"
condition_7 = "(doggo=='None' and floofer=='None' and pupper=='None' and puppo=='puppo')"

arr = [condition_1, condition_2, condition_3, condition_4, condition_5, condition_6, condition_7]

condition = " or ".join(arr)

rows_to_remove = df_wrd_twitter.query(condition)

rows_to_remove['tweet_id'].count()

0

In [78]:
df_wrd_twitter.query("name=='None'")['tweet_id'].count()

0

<a id='analyzing'></a>
# IV. Analyzing Data 
---

<a id='visualizing'></a>
# V. Visualizing Data 
---